### PR TITLE
many: fix travis; fix native and cross-compiling

### PR DESCRIPTION
**Pull request by anonymouse64**
This PR fixes cross-compiling the gadget snap, which now works properly everywhere, on LP, locally with lxd or multipass on native arches, or also cross-compiled with lxd/multipass, etc. Snapcraft 4.0 no longer supports --target-arch, so we no longer use that and this snapcraft.yaml now correctly builds either from amd64 to armhf or from armhf to armhf with supported snapcraft features in the architectures key of the snapcraft.yaml.

Building locally from this will also work if the user added the snappy-dev image PPA manually with --destructive-mode too.

### CODE REVIEW

**Comment by anonymouse64** [comment]
I will port this to the 20-arm64 branch later.

**Review by xnox — CHANGES_REQUESTED** [review]
Why do current snappy-dev PPA arm64 builds work, and armhf ones don't?

What is the smallest diff / change to unbreak that, as a snap build in launchpad? Note launchpad does setup etc/apt/sources and uses destructive mode to build snaps.

Have you created a snap in launchpad with these changes, and can you point at build logs?

These changes are a lot more than what I was expecting.

**Comment by anonymouse64** [comment]
> Have you created a snap in launchpad with these changes, and can you point at build logs?

Yes, see https://launchpad.net/~anonymouse67/+snap/pi-gadget-snap and specifically build log: https://launchpad.net/~anonymouse67/+snap/pi-gadget-snap/+build/930577 for armhf native build on LP. 

For the other use cases, here's the log from a focal install on an armhf device where the user manually adds the snappy-dev PPA: https://pastebin.ubuntu.com/p/xZT6fMXXmM/.

**Comment by anonymouse64** [comment]
Also note that this PR will fix the travis CI checks to the extent that they are actually checking that the builds work on focal rather than bionic as they were previously. The travis job is still somewhat wrong because it's not using the snappy-dev PPA, but that will be fixed in the follow-up that is currently blocked on waiting for snapcraft 4.0 to be fixed as per that PR.

**Comment by anonymouse64** [comment]
I decided to update this to just use the package-repositories key to build the snap and not do that in a different PR, so this now properly does cross-compilation too.

It also fixes travis to work with snapcraft 4.0 because with 4.0 you can't run snapcraft with sudo.

**Comment by anonymouse64** [comment]
For lp build example see https://launchpad.net/~anonymouse67/+snap/pi-gadget-snap/+build/992153

**Comment by anonymouse64** [comment]
gentle ping @xnox and @waveform80

**Comment by knitzsche** [comment]
I just built a gadget from https://github.com/anonymouse64/pi-gadget/tree/bugfix/uc20-backport-arm64-to-armhf and it seemd to work fine in a uc20 pi image. Got through console-conf to shell on boot.
I like the conservation of cross compile (amd64 - > armhf in this case) so that I can build the gadget on amd64 workstation.

File: snapcraft.yaml
Comment (xnox) [inline snapcraft.yaml]: Is this branch targetting only arm64, or only armhf, or both?

Cause above triplet is set to armhf, yet only arm64 cross-compiler is installed. Unless things that are build here need compilation for armhf tripplet irrespective of armhf/arm64 (freestanding?! but then it would not run on a subset of AArch64 cpus without the optional AArch32 instruction set)

File: snapcraft.yaml
Comment (xnox) [inline snapcraft.yaml]: Why is this if block needed? Surely both in native, and cross case, BUILD_ARCH_TRIPLET must have the same value, no?

File: snapcraft.yaml
Comment (xnox) [inline snapcraft.yaml]: The sources.list is arch qualified, and is correct for either native or cross compilation, does it need to be in a conditional?

File: snapcraft.yaml
Comment (anonymouse64) [reply snapcraft.yaml]: ah thanks for catching this, this is actually a problem, since this PR is targeting the armhf branch, this should be

```suggestion
          - libc6-dev-armhf-cross:amd64
```
(or whatever the actual packages are called)

but to answer your question, the `try` is always executed on all architectures, but it's not treated as fatal if snapcraft doesn't find this package available when it is building. This is because the `on ... to ...` support in snapcraft is broken because --target-arch is what sets the arch in `to ...` but we can't set that anymore except when building dangerously.

File: snapcraft.yaml
Comment (anonymouse64) [reply snapcraft.yaml]: fair we probably can just unconditionally set BUILD_ARCH_TRIPLET like this, I will test this out

**Comment by tobias-bischoff** [comment]
I've been using this in my fork of this gadget snap, and have had no issues so far. Tested with multiple Raspberry Pi models: 3B, 3B+, 4B. Are there any issues that speak against merging this? Upstream integration is always better than patches in git stashes 😄 Thanks in advance!

**Comment by anonymouse64** [comment]
@tobias-grasse sorry this is blocked waiting for me to address @xnox's feedback, I will do so soon hopefully now that snapd 2.48 is mostly out the door

File: snapcraft.yaml
Comment (anonymouse64) [reply snapcraft.yaml]: yes you're right it can always be used

File: snapcraft.yaml
Comment (anonymouse64) [reply snapcraft.yaml]: this works to set it unconditionally

File: snapcraft.yaml
Comment (anonymouse64) [reply snapcraft.yaml]: fixed this

**Comment by anonymouse64** [comment]
Alright so this branch has been updated to address @xnox's feedback, I force-pushed it to rebase on top of 20-armhf upstream here. I tested building it with the following methods:
* [x] with an amd64 Launchpad snap recipe that builds out of the snappy-dev/image PPA, see this link for an example build: https://code.launchpad.net/~anonymouse67/+snap/pi-gadget-snap/+build/1255460 [1]
* [x] locally on an amd64 system using LXD provider for snapcraft (thus cross-compiling)
* [x] natively on a raspberry pi 4 running Ubuntu Server 20.04

Thus, this should be ready now, but note that we should probably hold off on merging it for just a little bit more, since in order to cross-compile this gadget properly with i.e. LXD or multipass and vanilla snapcraft, you still need to pass `--enable-experimental-package-repositories`, but that makes the snap not buildable on LP, as LP doesn't yet know how to handle additional snapcraft arguments when trying to build a git repo / snap recipe. 

What we could do is I could revert that patch adding `package-repositories` and we merge this as-is, and open a followup just adding package-repositories which we don't merge until snapcraft has landed that out of experimental. I spoke with the snapcraft team today about package-repositories and they are apparently very close to getting that out of experimental for the next release, which is due at the end of this week, so maybe just waiting for this PR is fine.

In any case, I would like to get this PR +1d again before porting the relevant patches to the arm64 branch. 

[1] I tested building in LP by commenting out the package-repositories bit in the snapcraft.yaml

**Review by xnox — APPROVED** [review]

**Pull request merged**
